dotnet/WPF/VisualLayerIntegration/VisualLayerIntegration/BarGraphUtility/BarGraph.cs

```diff
@@ -0,0 +1,479 @@
+﻿//  ---------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+// 
+//  The MIT License (MIT)
+// 
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+// 
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+// 
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//  ---------------------------------------------------------------------------------
+
+using SharpDX;
+using SharpDX.DirectWrite;
+using SharpDX.Direct2D1;
+using TextAntialiasMode = SharpDX.Direct2D1.TextAntialiasMode;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using SysWin = System.Windows;
+using Windows.UI;
+using Windows.UI.Composition;
+
+namespace BarGraphUtility
+{
+    class BarGraph
+    {
+        private float[] graphData;
+
+        private Compositor compositor;
+        private IntPtr hwnd;
+
+        private float graphWidth, graphHeight;
+        private float shapeGraphContainerHeight, shapeGraphContainerWidth, shapeGraphOffsetY, shapeGraphOffsetX;
+        private float barWidth, barSpacing;
+        private float maxBarValue;
+
+        private GraphBarStyle graphBarStyle;
+        private List<Windows.UI.Color> graphBarColors;
+
+        private Hashtable barValueMap;
```

Comment:
Hashtable pretty much only exists for very old code. You should use the strongly typed Dictionary<K,V> in its place.

Reply:
updated